src/agents/common/utils.py

```diff
@@ -47,6 +51,44 @@ def filter_messages(
     return filtered
 
 
+def filter_valid_messages(
+    messages: Sequence[BaseMessage] | list[BaseMessage],
+) -> Sequence[BaseMessage]:
+    """
+    Filters the invalid sequence of messages.
+    For example:
+    - An assistant message with 'tool_calls' must be followed by tool messages responding to each 'tool_call_id'.
+    - A tool message must be preceded by an assistant message with 'tool_calls'.
+    This method should be used when the LLMs are being invoked with messages.
+    """
+
+    filtered: list[BaseMessage] = []
+    for i, message in enumerate(messages):
+        if isinstance(message, ToolMessage):
+            # ToolMessage should not be added directly
+            continue
+        elif isinstance(message, AIMessage) and message.tool_calls:
+            # check if the next messages are tool calls as requested by AIMessage.
+            is_valid_sequence = True
+            for j, _ in enumerate(message.tool_calls):
+                next_message_index = i + j + 1  # +1 because the index starts from zero.
+                if next_message_index >= len(messages) or not isinstance(
+                    messages[next_message_index], ToolMessage
+                ):
+                    is_valid_sequence = False
+                    break
+            # if the sequence of AIMessage and ToolMessage(s) is completed, then append the completed sequence.
+            if is_valid_sequence:
+                # append the AIMessage and its corresponding ToolMessages.
+                filtered.append(message)
+                for j, _ in enumerate(message.tool_calls):
+                    filtered.append(messages[i + j + 1])
+        else:
+            # append other valid messages.
+            filtered.append(message)
+    return filtered
```

Comment:
You could avoid all the nested `for` loops here with something like
```suggestion
filtered: list[BaseMessage] = []
    i = 0

    while i < len(messages):
        message = messages[i]

        if isinstance(message, ToolMessage):
            # ToolMessage should not be added directly.
            i += 1
            continue

        if isinstance(message, AIMessage) and message.tool_calls:
            # Check if the next messages are tool calls as requested by AIMessage.
            tool_call_count = len(message.tool_calls)
            next_messages = messages[i + 1 : i + 1 + tool_call_count]

            if len(next_messages) == tool_call_count and all(
                isinstance(next_msg, ToolMessage) for next_msg in next_messages
            ):
                # Append the AIMessage and its corresponding ToolMessages.
                filtered.append(message)
                filtered.extend(next_messages)
                i += tool_call_count  # Skip the tool messages

        else:
            # Append other valid messages.
            filtered.append(message)

        i += 1

    return filtered
```

Reply:
updated

Reply:
Nice, thanks. 👍🏻